keras/backend/tensorflow/numpy.py

```diff
@@ -10,17 +9,17 @@
 from keras.backend import config
 from keras.backend import standardize_dtype
 from keras.backend.common import dtypes
+from keras.backend.tensorflow import sparse
 from keras.backend.tensorflow.core import convert_to_tensor
 
 
+@sparse.element_wise_binary_on_union(tf.sparse.add)
```

Comment:
Let's apply the following replacements:

`element_wise` -> `elementwise` (everywhere)
`on_union` -> just `union` (to be consistent with `intersection`)

Reply:
Question: is this new system really better than `tf.sparse.add`? Does it have performance implications? The previous "`if` + use `tf.sparse`" setup was very readable and easy to maintain. The new system is fairly high in abstraction and complexity. Are the benefits commensurate?

Reply:
Will do the replacements

> is this new system really better than `tf.sparse.add`?

So there are 2 times 3 cases: `SparseTensor` / `IndexedSlices` and **union** (add, maximum...), **intersection** (multiply), **division** (divide, true_divide, mod, floor_divide). `tf.sparse` happens to mostly support the "union" operators out of the box, so it was very little code. But it didn't support the other use cases and `IndexedSlices` is not supported for any operation. There is a lot of code for the 5 other use cases (for instance, this CL moved the long chunk of code to get `multiply` working for sparse from `numpy.py` to `sparse.py`).

One thing we could do differently is have the `elementwise_binary_union` decorator only add `IndexedSlices` support and not `SparseTensor`s to have the `tf.sparse.add` code inline. I initially had it this way, but then I decided to make it look consistent.

> Does it have performance implications?

For `tf.sparse.add` specifically, the only performance impact is the call to the decorator, which we need anyway for the `IndexedSlices` support. It does use `tf.sparse.add` directly, this is not re-implemented.

> The new system is fairly high in abstraction and complexity.

I agree. After all was done to handle all the possible combinations, it was a lot more complex than initially expected.

> Are the benefits commensurate?

Well the benefit is that with one decorator, you make an op `SparseTensor` and `IndexedSlices` compatible. And it's self documenting, you know which ops support `SparseTensor` and `IndexedSlices` (we could generate documentation based on the decorator). I could have decorated more ops actually (e.g. logical operations).

I want to look at JAX and see what approach makes sense for JAX. Maybe it will be similar, maybe it will make me rethink this whole thing.

Reply:
Thanks for the clarifications, this all sounds reasonable. Please let me know what you think about the JAX situation, and then we can finalize the design.

Reply:
Thanks!

One other thing is that implementing support in the ops directly adds complexity. For instance, in practice, this union / intersection business is never needed for `IndexedSlices`. In optimizers the only times `IndexedSlices` are combined, they always have the same indices. But in the ops, we can't just make this assumption.